src/Microsoft.OData.Core/Json/JsonReader.cs

```diff
@@ -921,59 +927,55 @@ private string ConsumeTokenToString(int characterCount)
+
             // We need to make sure we have more room in the buffer to read characters into
             if (this.storedCharacterCount == this.characterBuffer.Length)
             {
                 // No more room in the buffer, move or grow the buffer.
                 if (this.tokenStartIndex == this.storedCharacterCount)
                 {
-                    // If the buffer is empty (all characters were consumed from it)
-                    // just start over.
+                    // If the buffer is empty (all characters were consumed from it), just start over.
                     this.tokenStartIndex = 0;
                     this.storedCharacterCount = 0;
                 }
-                else if (this.tokenStartIndex > (this.characterBuffer.Length - MaxCharacterCountToMove))
-                {
-                    // Some characters were consumed, we can just move them in the buffer
-                    // to get more room without allocating.
-                    Array.Copy(
-                        this.characterBuffer,
-                        this.tokenStartIndex,
-                        this.characterBuffer,
-                        0,
-                        this.storedCharacterCount - this.tokenStartIndex);
-                    this.storedCharacterCount -= this.tokenStartIndex;
-                    this.tokenStartIndex = 0;
-                }
-                else
+                else if (this.tokenStartIndex < this.characterBuffer.Length / 2)
                 {
                     // The entire buffer is full of unconsumed characters
-                    // We need to grow the buffer.
-                    // Double the size of the buffer.
+                    // We need to grow the buffer. Double the size of the buffer.
                     int newBufferSize = this.characterBuffer.Length * 2;
-                    char[] newCharacterBuffer = new char[newBufferSize];
+                    newBufferSize = newBufferSize < 0 ? int.MaxValue : newBufferSize; // maybe overflow
+
+                    char[] newCharacterBuffer = BufferUtils.RentFromBuffer(ArrayPool, newBufferSize);
 
                     // Copy the existing characters to the new buffer.
-                    Array.Copy(
-                        this.characterBuffer,
-                        0,  // The tokenStartIndex is 0 - we checked above.
-                        newCharacterBuffer,
-                        0,
-                        this.characterBuffer.Length);  // The storedCharacterCount is the size of the old buffer
+                    Array.Copy(this.characterBuffer, this.tokenStartIndex, newCharacterBuffer, 0,
+                        this.storedCharacterCount - this.tokenStartIndex);
 
                     // And switch the buffers
+                    BufferUtils.ReturnToBuffer(ArrayPool, this.characterBuffer);
                     this.characterBuffer = newCharacterBuffer;

```

Comment:
>; [](start = 61, length = 1)

Don't we also need to reset tokenStartIndex and storedCharacterCount?

Reply:
Yes. I fixed that.

---
In reply to: [253987895](https://github.com/OData/odata.net/pull/1388#discussion_r253987895) [](ancestors = 253987895)